effekt/shared/src/main/scala/effekt/RecursiveDescent.scala

```diff
@@ -466,53 +484,62 @@ class RecursiveDescent(positions: Positions, tokens: Seq[Token], source: Source)
   def effectDef(): Def =
     nonterminal:
       // effect <NAME> = <EFFECTS>
-      EffectDef(`effect` ~> idDef(), maybeTypeParams(), `=` ~> effects())
+      documented { doc =>
+        EffectDef(`effect` ~> idDef(), maybeTypeParams(), `=` ~> effects(), doc)
+      }
 
   // effect <NAME>[...](...): ...
   def operationDef(): Def =
     nonterminal:
-      `effect` ~> operation() match {
-        case op @ Operation(id, tps, vps, bps, ret) =>
-          InterfaceDef(IdDef(id.name) withPositionOf op, tps, List(Operation(id, Nil, vps, bps, ret) withPositionOf op))
+      val doc = maybeDocumentation()
+      `effect` ~> operation(doc) match {
+        case op @ Operation(id, tps, vps, bps, ret, doc) => // TODO: should the same `doc` be used here?
+          InterfaceDef(IdDef(id.name) withPositionOf op, tps, List(Operation(id, Nil, vps, bps, ret, doc) withPositionOf op), doc)
       }
 
-  def operation(): Operation =
+  def operation(doc: Doc): Operation =
     nonterminal:
       idDef() ~ params() ~ returnAnnotation() match {
-        case id ~ (tps, vps, bps) ~ ret => Operation(id, tps, vps, bps, ret)
+        case id ~ (tps, vps, bps) ~ ret => Operation(id, tps, vps, bps, ret, doc)
       }
 
   def interfaceDef(): InterfaceDef =
     nonterminal:
-      InterfaceDef(`interface` ~> idDef(), maybeTypeParams(), `{` ~> manyWhile(`def` ~> operation(), `def`) <~ `}`)
+      documented { doc =>
+        InterfaceDef(`interface` ~> idDef(), maybeTypeParams(),
+          `{` ~> manyWhile(documented { opDoc => `def` ~> operation(opDoc) }, documentedKind == `def`) <~ `}`, doc)
+      }
 
   def namespaceDef(): Def =
     nonterminal:
+      val doc = maybeDocumentation()
       consume(`namespace`)
       val id = idDef()
       // namespace foo { <DEFINITION>* }
-      if peek(`{`) then braces { NamespaceDef(id, definitions()) }
+      if peek(`{`) then braces { NamespaceDef(id, definitions(), doc) }
       // namespace foo
       // <DEFINITION>*
-      else { semi(); NamespaceDef(id, definitions()) }
+      else { semi(); NamespaceDef(id, definitions(), doc) }
 
 
   def externDef(): Def =
     nonterminal:
-      { peek(`extern`); peek(1).kind } match {
-        case `type`      => externType()
-        case `interface` => externInterface()
-        case `resource`  => externResource()
-        case `include`   => externInclude()
+      val doc = maybeDocumentation()
+      consume(`extern`)
+      peek.kind match {
+        case `type`      => externType(doc)
+        case `interface` => externInterface(doc)
+        case `resource`  => externResource(doc)
+        case `include`   => externInclude(doc)
         // extern """..."""
-        case s: Str      => externString()
+        case s: Str      => externString(doc)
         case Ident(_) | `pure` =>
           // extern IDENT def ...
-          if (peek(2, `def`)) externFun()
+          if (peek(1, `def`)) externFun(doc)
```

Comment:
Why do we need to consume the extern here upfront? This deviates from the usual style that nonterminals consume their tokens.

Reply:
I thought it makes more sense this way, other definitions above (e.g. `namespaceDef`) are similar in that we first consume the documentation, then consume some identifier and then continue based on the next symbols. Otherwise we need to add a separate `maybeDocumentation()` call to all `extern...` parsers.

I see the resulting deviation in `extern...` though, so should I change it?